receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go

```diff
@@ -0,0 +1,148 @@
+)
+
+const (
+	socketPath        = "/var/lib/kubelet/pod-resources/kubelet.sock"
+	connectionTimeout = 10 * time.Second
+	taskTimeout       = 10 * time.Second
+)
+
+var (
+	instance *PodResourcesStore
+	once     sync.Once
+)
+
+type ContainerInfo struct {
+	podName       string
+	containerName string
+	namespace     string
+}
+
+type ResourceInfo struct {
+	resourceName string
+	deviceID     string
+}
+
+type PodResourcesClientInterface interface {
+	ListPods() (*v1.ListPodResourcesResponse, error)
+}
+
+type PodResourcesStore struct {
+	containerInfoToResourcesMap map[ContainerInfo][]ResourceInfo
+	resourceToPodContainerMap   map[ResourceInfo]ContainerInfo
+	lastRefreshed               time.Time
+	ctx                         context.Context
+	cancel                      context.CancelFunc
+	logger                      *zap.Logger
+	podResourcesClient          PodResourcesClientInterface
+}
+
+func NewPodResourcesStore(logger *zap.Logger) *PodResourcesStore {
+	once.Do(func() {
+		podResourcesClient, _ := kubeletutil.NewPodResourcesClient()
+		ctx, cancel := context.WithCancel(context.Background())
+		instance = &PodResourcesStore{
+			containerInfoToResourcesMap: make(map[ContainerInfo][]ResourceInfo),
+			resourceToPodContainerMap:   make(map[ResourceInfo]ContainerInfo),
+			lastRefreshed:               time.Now(),
+			ctx:                         ctx,
+			cancel:                      cancel,
+			logger:                      logger,
+			podResourcesClient:          podResourcesClient,
+		}
+
+		go func() {
+			refreshTicker := time.NewTicker(time.Second)
+			for {
+				select {
+				case <-refreshTicker.C:
+					instance.refreshTick()
+				case <-instance.ctx.Done():
+					refreshTicker.Stop()
+					return
+				}
+			}
+		}()
+	})
+	return instance
+}
+
+func (p *PodResourcesStore) refreshTick() {
+	now := time.Now()
+	if now.Sub(p.lastRefreshed) >= taskTimeout {
+		p.refresh()
+		p.lastRefreshed = now
+	}
+}
+
+func (p *PodResourcesStore) refresh() {
+	doRefresh := func() {
+		p.updateMaps()
+	}
+
+	refreshWithTimeout(p.ctx, doRefresh, taskTimeout)
+}
+
+func (p *PodResourcesStore) updateMaps() {
+	p.containerInfoToResourcesMap = make(map[ContainerInfo][]ResourceInfo)
+	p.resourceToPodContainerMap = make(map[ResourceInfo]ContainerInfo)
+
+	devicePods, err := p.podResourcesClient.ListPods()
+
+	if err != nil {
+		p.logger.Error(fmt.Sprintf("Error getting pod resources: %v", err))
+		return
+	}
+
+	for _, pod := range devicePods.GetPodResources() {
+		for _, container := range pod.GetContainers() {
+			for _, device := range container.GetDevices() {
+
+				containerInfo := ContainerInfo{
+					podName:       pod.GetName(),
+					namespace:     pod.GetNamespace(),
+					containerName: container.GetName(),
+				}
+
+				for _, deviceID := range device.GetDeviceIds() {
+					resourceInfo := ResourceInfo{
+						resourceName: device.GetResourceName(),
```

Comment:
We might want to pass in a set of device resource names that we are interested in, so that we don't waste space storing *all* devices.

Reply:
would it make sense to have a add_resource_name method which adds that to a set which we can then use for lookup and populate the map in the next run?

Reply:
Yes I like that.